core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java

```diff
@@ -814,9 +830,24 @@ public void close()
                 return;
             }
             finished = true;
-            if (exchangeSource != null) {
-                exchangeSource.close();
-            }
+            addCallback(exchangeSourceFuture, new FutureCallback<>()
+            {
+                @Override
+                public void onSuccess(ExchangeSource exchangeSource)
+                {
+                    try {
+                        exchangeSource.close();
+                    }
+                    catch (RuntimeException e) {
+                        log.warn(e, "error closing exchange source");
+                    }
+                }
+
+                @Override
+                public void onFailure(Throwable ignored)
```

Comment:
why? no log?

Reply:
If the `ExchangeSource` creation fails it will be reported through the `getNext`. This callback is used only to close `ExchangeSource` safely.

Reply:
maybe add comment in code